core/src/test/java/org/apache/calcite/test/enumerable/EnumerableMergeUnionTest.java

```diff
@@ -0,0 +1,282 @@
+            + "      EnumerableTableScan(table=[[s, emps]])\n")
+        .returnsOrdered(
+            "empid=1; name=Bill",
+            "empid=6; name=Guy",
+            "empid=10; name=Gabriel",
+            "empid=12; name=Paul",
+            "empid=29; name=Anibal",
+            "empid=40; name=Emmanuel",
+            "empid=45; name=Pascal");
+  }
+
+  @Test void mergeUnionAllOrderByName() {
+    tester(false,
+        new JdbcTest.HrSchemaBig(),
+        "select * from (select empid, name from emps where name like 'G%' union all select empid, name from emps where name like '%l') order by name")
+        .explainContains("EnumerableMergeUnion(all=[true])\n"
+            + "  EnumerableSort(sort0=[$1], dir0=[ASC])\n"
+            + "    EnumerableCalc(expr#0..4=[{inputs}], expr#5=['G%'], expr#6=[LIKE($t2, $t5)], empid=[$t0], name=[$t2], $condition=[$t6])\n"
+            + "      EnumerableTableScan(table=[[s, emps]])\n"
+            + "  EnumerableSort(sort0=[$1], dir0=[ASC])\n"
+            + "    EnumerableCalc(expr#0..4=[{inputs}], expr#5=['%l'], expr#6=[LIKE($t2, $t5)], empid=[$t0], name=[$t2], $condition=[$t6])\n"
+            + "      EnumerableTableScan(table=[[s, emps]])\n")
+        .returnsOrdered(
+            "empid=29; name=Anibal",
+            "empid=1; name=Bill",
+            "empid=40; name=Emmanuel",
+            "empid=10; name=Gabriel",
+            "empid=10; name=Gabriel",
+            "empid=6; name=Guy",
+            "empid=45; name=Pascal",
+            "empid=12; name=Paul");
+  }
+
+  @Test void mergeUnionOrderByName() {
+    tester(false,
+        new JdbcTest.HrSchemaBig(),
+        "select * from (select empid, name from emps where name like 'G%' union select empid, name from emps where name like '%l') order by name")
+        .explainContains("EnumerableMergeUnion(all=[false])\n"
+            + "  EnumerableSort(sort0=[$1], dir0=[ASC])\n"
+            + "    EnumerableCalc(expr#0..4=[{inputs}], expr#5=['G%'], expr#6=[LIKE($t2, $t5)], empid=[$t0], name=[$t2], $condition=[$t6])\n"
+            + "      EnumerableTableScan(table=[[s, emps]])\n"
+            + "  EnumerableSort(sort0=[$1], dir0=[ASC])\n"
+            + "    EnumerableCalc(expr#0..4=[{inputs}], expr#5=['%l'], expr#6=[LIKE($t2, $t5)], empid=[$t0], name=[$t2], $condition=[$t6])\n"
+            + "      EnumerableTableScan(table=[[s, emps]])\n")
+        .returnsOrdered(
+            "empid=29; name=Anibal",
+            "empid=1; name=Bill",
+            "empid=40; name=Emmanuel",
+            "empid=10; name=Gabriel",
+            "empid=6; name=Guy",
+            "empid=45; name=Pascal",
+            "empid=12; name=Paul");
+  }
+
+  @Test void mergeUnionOrderByNameWithLimit() {
+    tester(false,
+        new JdbcTest.HrSchemaBig(),
+        "select * from (select empid, name from emps where name like 'G%' union select empid, name from emps where name like '%l') order by name limit 3")
+        .explainContains("EnumerableLimit(fetch=[3])\n"
```

Comment:
Can you please propagate limit to the downstream nodes?

In other words, if you want 3 rows from the union, then you could limit each branch with 3 so they can use limited sort rather than sorting of the full result.

WDYT?

Reply:
That would work, correct me if I am wrong, only if we have a "pure LIMIT". If we have a LIMIT + OFFSET (or just OFFSET), it cannot be pushed into the downstream nodes.

Reply:
`OFFSET x LIMIT y` can be transformed to the downstream `LIMIT x+y`

Reply:
Done.

Reply:
This is a good point!